### ISSUE TITLE
Implementing `Copy` can be a breaking change

### ISSUE THREAD

**Issue by Jules-Bertholet** [rust-lang/rust#126179]
I tried this code:

```rust
#[derive(Clone)]
struct Foo<'a>(&'a ());

// Uncomment this line and the code stops compiling
//impl Copy for Foo<'static> {}

struct Bar<'a>(Foo<'a>);

fn do_thing<'a>(b: Bar<'a>) {
    drop(b.0);
}
```

I expected to see this happen: Uncommenting the line in question does not break the code, as implementing `Copy` should not be a breaking change

Instead, this happened:

```
error: lifetime may not live long enough
  --> src/lib.rs:10:10
   |
9  | fn do_thing<'a>(b: Bar<'a>) {
   |             -- lifetime `'a` defined here
10 |     drop(b.0);
   |          ^^^ copying this value requires that `'a` must outlive `'static`
```

### Meta

Regressed in 1.42 according to Godbolt.

@rustbot label A-borrow-checker A-lifetimes A-traits T-types regression-from-stable-to-stable

**Comment by QuineDot** [comment]
[Intentionally denied since Rust 1.41.1.](https://blog.rust-lang.org/2020/02/27/Rust-1.41.1.html#respecting-a-static-lifetime-in-a-copy-implementation)

[Here's the PR.](https://github.com/rust-lang/rust/pull/69145)

**Comment by Jules-Bertholet** [comment]
> [Intentionally denied since Rust 1.41.1.](https://blog.rust-lang.org/2020/02/27/Rust-1.41.1.html#respecting-a-static-lifetime-in-a-copy-implementation)

The example from the blogpost has *two* `drop`s, which of course must require the field type to be `Copy` because otherwise you would access a field that has already been moved out of. My example has only one `drop`, which should not require any such restriction. So it seems the fix PR was overly strict.

**Comment by Nilstrieb** [comment]
Adding a copy impl for a specific lifetime is not something that should ever be done, and a bug in the program.

**Comment by BoxyUwU** [comment]
Why was this marked regression stable-to-stable?

**Comment by Jules-Bertholet** [comment]
> Removing the regression labels since this was intentional

Looking through discussion on the PR and on Zulip, I haven't found any evidence that this is the case.

**Comment by Jules-Bertholet** [comment]
> Adding a copy impl for a specific lifetime is […] a bug in the program.

That's perhaps a reasonable position to take, but it's not one the language team has ever FCPed AFAIK.

**Comment by Nilstrieb** [comment]
I don't think we need a lang FCP to decide that implementing copy for only some lifetimes is nonsense^^

**Comment by Jules-Bertholet** [comment]
The documentation of [`Copy`](https://doc.rust-lang.org/std/marker/trait.Copy.html) does not say that. Implementing `Copy` for only some lifetimes may not seem very useful. But most programs in the space of all valid Rust programs are not useful, which does not make them any less valid; and just because you or I cannot think of an obvious use-case, that does not mean that none exists.

**Comment by traviscross** [comment]
Minimizing further:

```rust
#[derive(Clone)]
struct W<T>(T);
impl Copy for W<&'static ()> {}

fn test<'a>(x: W<&'a ()>) {
    let _x = x;
    //~^ ERROR lifetime may not live long enough
    //~| NOTE copying this value requires that `'a` must outlive `'static`
}
```

This is, I would imagine, a result of the fact that "lifetimes don't participate in trait selection".  E.g.

```rust
struct W<T>(T);

trait OnW {
    fn method(&self) {}
}

trait OnWStatic {
    fn method(&self) {}
}

impl<'a> OnW for W<&'a ()> {}
impl OnWStatic for &W<&'static ()> {}

fn test<'a>(x: W<&'a ()>) {
    (&&x).method();
    //~^ ERROR borrowed data escapes outside of function
    //~| NOTE `x` escapes the function body here
    //~| NOTE argument requires that `'a` must outlive `'static`
}
```

See [*Trait solving > Selection*](https://rustc-dev-guide.rust-lang.org/traits/resolution.html#selection) in the dev guide:

> Note that because of how lifetime inference works, it is not possible to give back immediate feedback as to whether a unification or subtype relationship between lifetimes holds or not. Therefore, lifetime matching is *not* considered during selection. This is reflected in the fact that subregion assignment is infallible. This may yield lifetime constraints that will later be found to be in error (in contrast, the non-lifetime-constraints have already been checked during selection and can never cause an error, though naturally they may lead to other errors downstream).

So while I wouldn't go so far as to say that implementing a trait for one particular lifetime is meaningless, it seems not likely to be usefully supported in Rust for the foreseeable future.

**Comment by apiraino** [comment]
Going to remove the prioritization label, by reading the comments I don't see an actionable here. But If this _is_ a regression please change my mind.

@rustbot label -I-prioritize -regression-from-stable-from-stable

**Comment by Jules-Bertholet** [comment]
My understanding is, this issue stems from the fact that moving vs copying a non-`Drop` type compiles to different MIR, and this choice must be made without lifetime information. However, according to today's Miri, these operations are identical from an opsem perspective (https://github.com/rust-lang/unsafe-code-guidelines/issues/188); so theoretically, we could represent them with the same MIR, and have the borrow checker do a little more work to verify that no use-after-move is present. Is this correct?
